src/repositories/gitlab.rs

```diff
@@ -0,0 +1,222 @@
+            }),
+        )?;
+        if let Some(repo) = response.data.and_then(|d| d.project) {
+            Ok(Some(Repository {
+                id: repo.id,
+                name_with_owner: repo.name_with_namespace,
+                description: repo.description,
+                last_activity_at: repo.last_activity_at,
+                stars: repo.star_count,
+                forks: repo.forks_count,
+                issues: repo.open_issues_count.unwrap_or(0),
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn fetch_repositories(&self, ids: &[String]) -> Result<FetchRepositoriesResult> {
+        let response: GraphResponse<GraphProjects<Option<GraphProject>>> = self.graphql(
+            GRAPHQL_UPDATE,
+            serde_json::json!({
+                "ids": ids,
+            }),
+        )?;
+        let mut ret = FetchRepositoriesResult::default();
+        // When gitlab doesn't find an ID, it simply doesn't list it. So we need to actually check
+        // which nodes remain at the end to delete their DB entry.
+        let mut node_ids: HashSet<&String> = ids.iter().collect();
+
+        if let Some(data) = response.data {
+            if !response.errors.is_empty() {
+                failure::bail!("error updating repositories: {:?}", response.errors);
+            }
+            for node in data.projects.nodes.into_iter() {
+                if let Some(node) = node {
+                    let repo = Repository {
+                        id: node.id,
+                        name_with_owner: node.name_with_namespace,
+                        description: node.description,
+                        last_activity_at: node.last_activity_at,
+                        stars: node.star_count,
+                        forks: node.forks_count,
+                        issues: node.open_issues_count.unwrap_or(0),
+                    };
+                    let id = repo.id.clone();
+                    node_ids.remove(&id);
+                    ret.present.insert(id, repo);
+                }
+            }
+
+            // Those nodes were not returned by gitlab, meaning they don't exist (anymore?).
+            ret.missing = node_ids.into_iter().map(|s| s.to_owned()).collect();
+
+            Ok(ret)
+        } else {
+            failure::bail!("no data")
+        }
+    }
+}
+
+impl GitLab {
+    fn graphql<T: serde::de::DeserializeOwned + std::fmt::Debug>(
+        &self,
+        query: &str,
+        variables: impl serde::Serialize,
+    ) -> Result<GraphResponse<T>> {
+        Ok(self
+            .client
+            .post(&format!("https://{}/api/graphql", self.host))
+            .json(&serde_json::json!({
+                "query": query,
+                "variables": variables,
+            }))
+            .send()?
+            .error_for_status()?
+            .json()?)
+    }
+}
+
+#[derive(Debug, failure::Fail)]
+#[fail(display = "rate limit reached")]
+struct RateLimitReached;
```

Comment:
This type is not used. We should detect the rate limit on GitLab's side, and share the same error type we use for GitHub.